aten/src/ATen/native/cuda/DistributionTemplates.h

```diff
@@ -32,33 +32,31 @@ namespace {
 // launch bounds used for kernels utilizing TensorIterator
 const uint32_t block_size_bound = 256;
 const uint32_t grid_size_bound = 4;
-// number of randoms given by distributions like curand_uniform4, curand_uniform2_double
-// used in calculating philox offset.
-const uint32_t curand4_engine_calls = 4;
+// At the time of writing, there is no curand_* call that increments the offset by more than 4.
+// See: https://docs.nvidia.com/cuda/archive/11.8.0/curand/group__DEVICE.html
+const uint32_t max_generator_offsets_per_curand_call = 4;
 
 // utility function that calculates proper philox_offset
 // for distributions utilizing TensorIterator. For distributions using
 // TensorIterator, we are using a grid-stride loop with each
 // thread yielding one element per thread. For the edge of the grid-stride
 // loop, if the tensor size is large, the unroll loop will kick in and the float4
 // from curand4 will start getting utilized (for common tensor sizes, we end up
-// using rand.x from each thread). Hence, the philox_offset is
-// (number of elements per thread * number of engine calls), which makes
+// using rand.x from each thread). The philox_offset calculation was changed to
+// (number of elements per thread * maximum generator increment per "curand_*" call), which makes
 // sure that philox offset increment is not less than the number of randoms used
 // in each thread.
-std::tuple<uint64_t, dim3, dim3> calc_execution_policy(int64_t total_elements) {
+std::tuple<uint64_t, dim3, dim3> calc_execution_policy(const int64_t total_elements, const uint32_t unroll_factor) {
   const uint64_t numel = static_cast<uint64_t>(total_elements);
   const uint32_t block_size = block_size_bound;
-  const uint32_t unroll = curand4_engine_calls;
   dim3 dim_block(block_size);
   dim3 grid((numel + block_size - 1) / block_size);
   uint32_t blocks_per_sm = at::cuda::getCurrentDeviceProperties()->maxThreadsPerMultiProcessor / block_size;
   grid.x = std::min(
       static_cast<uint32_t>(at::cuda::getCurrentDeviceProperties()->multiProcessorCount) * blocks_per_sm,
       grid.x);
   //number of times random will be generated per thread, to offset philox counter in thc random state
-  uint64_t counter_offset = ((numel - 1) / (block_size * grid.x * unroll) + 1)
-                                * curand4_engine_calls;
+  uint64_t counter_offset = ((numel - 1) / (block_size * grid.x * unroll_factor) + 1) * max_generator_offsets_per_curand_call;
```

Comment:
Does this `philox_offset` change impact the output of previously correct generation? e.g., after this change would it be expected to get the same RNG output for a deterministic generator via `torch.manual_seed`?

Reply:
It should only affect the previously wrong generations. I superficially checked my build vs. `torch=2.3.0` and some differences are (for CUDA only):
-  `torch.rand/torch.randn` calls with `dtype=torch.double`
-  `torch.randint()` calls for `dtype=int64, double, float, bfloat16` when `to-from` range is above 2**32 (`float` and `bfloat16` would have precision limitations in that range anyway). 
- `RReLU` with `dtype=torch.double`

Basically every call of `distribution_nullary_kernel` with `curand4_engine_calls/2` will have different results now. Should we add a `UserWarning` in those cases?